File: 11-Martin_Wattenberg/code.py
import tkinter as tk
import random
import threading
import time


class Point:
    """Class to represent a point in 2D space."""
    def __init__(self, x, y):
        self.x = x
        self.y = y


class ConnectApplet(tk.Canvas):
    def __init__(self, root, width, height):
        super().__init__(root, width=width, height=height, bg="white")
        self.pack()

        self.width = width
        self.height = height

        self.picture_image = tk.PhotoImage(width=width, height=height)
        self.create_image((0, 0), image=self.picture_image, anchor="nw")

        self.p = [
            Point(self.width // 2, 100),
            Point(self.width // 2 + 70, 170),
            Point(self.width // 2 - 70, 170),
        ]
        self.current = self.p[0]

        self.bind("<Button-1>", self.mouse_down)
        self.bind("<B1-Motion>", self.mouse_drag)

        self.running = True
        self.animation_thread = threading.Thread(target=self.run)
        self.animation_thread.start()

    def stop(self):
        """Stop the animation."""
        self.running = False
        self.animation_thread.join()

    def run(self):
        """Main animation thread."""
        while self.running:
            self.update_picture()
            self.repaint()
            time.sleep(0.02)

    def repaint(self):
        """Redraw the points on the canvas."""
        self.delete("all")
        self.create_image((0, 0), image=self.picture_image, anchor="nw")
        for point in self.p:
            self.create_oval(
                point.x - 5, point.y - 5, point.x + 5, point.y + 5,
                fill="red", outline="black"
            )

    def mouse_down(self, event):
        """Select the nearest point when clicked."""
        shortest = float("inf")
        for point in self.p:
            d2 = (point.x - event.x) ** 2 + (point.y - event.y) ** 2
            if d2 <= shortest:
                self.current = point
                shortest = d2

    def mouse_drag(self, event):
        """Move the selected point by dragging."""
        self.current.x = event.x
        self.current.y = event.y
        self.repaint()

    def update_picture(self):
        """Update the image with new colors."""
        for _ in range(5000):
            x = random.randint(0, self.width - 1)
            y = random.randint(0, self.height - 1)

            a1 = self.f(self.p[0], self.p[1], x, y)
            a2 = self.f(self.p[1], self.p[2], x, y)
            a3 = self.f(self.p[2], self.p[0], x, y)

            if a1 * a2 * a3 > 0:
                color = "#FFFFFF" if a1 * a2 > 0 and a2 * a3 > 0 else "#BEBEBE"
            else:
                color = "#000000"

            self.picture_image.put(color, (x, y))

    def f(self, p1, p2, x, y):
        """Calculate the determinant for the triangle."""
        return ((p1.x - x) * (p2.y - y) - (p1.y - y) * (p2.x - x)) // 13


def main():
    root = tk.Tk()
    root.title("ConnectApplet")
    applet = ConnectApplet(root, width=800, height=600)
    root.protocol("WM_DELETE_WINDOW", applet.stop)
    root.mainloop()


if __name__ == "__main__":
    main()

File: 11-Martin_Wattenberg/code.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;

public class ConnectApplet extends JPanel implements Runnable{

   Point current;
   Point[] p=new Point[3];

   Thread animation;
   int w,h;
   Image pictureImage;
   Graphics picture;

   public ConnectApplet() {
    w = getWidth();
    h = getHeight();
    pictureImage = createImage(w, h);
    picture = pictureImage.getGraphics();
    p[0] = new Point(w / 2, 100);
    p[1] = new Point(w / 2 + 70, 170);
    p[2] = new Point(w / 2 - 70, 170);
    current = p[0];
    addMouseListener(new MouseAdapter() {
        public void mousePressed(MouseEvent e) {
            mouseDown(e);
        }
    });
    addMouseMotionListener(new MouseAdapter() {
        public void mouseDragged(MouseEvent e) {
            mouseDrag(e);
        }
    });
}

   public void start() {

      animation=new Thread(this);

      animation.start();

   }

   public void stop() { animation=null; }

   public void run() {

      while (animation==Thread.currentThread()) {

         updatePicture();

         repaint();

         try {Thread.sleep(20);} catch (Exception e) {}

      }//v2/img/enterproject.gif

   }

   public void update(Graphics g) { paint(g); }

   public synchronized void paint(Graphics g) {

      g.drawImage(pictureImage,0,0,null);

   }

   // key: code in gray makes the program run.

   //      code in black draws the picture.



   public boolean mouseDown(MouseEvent e, int x, int y) {

      int shortest=Integer.MAX_VALUE;

      for (int i=0; i<p.length; i++) {

         int d2=(p[i].x-x)*(p[i].x-x)+

                (p[i].y-y)*(p[i].y-y);

          if (d2<=shortest) {

             current=p[i];

             shortest=d2;

          }

      }

      return mouseDrag(e,x,y);
   }
   public boolean mouseDrag(Event e, int x, int y) {
      current.x=x; current.y=y; return true;
   }
   synchronized void updatePicture() {
      for (int i=0; i<5000; i++) {
         int x=(int)(w*Math.random());
         int y=(int)(h*Math.random());
         long a1=f(p[0],p[1],x,y),
              a2=f(p[1],p[2],x,y),
              a3=f(p[2],p[0],x,y);
         picture.setColor(
            (int)(a1*a2*a3)>0 ? a1*a2>0 && a2*a3>0 ?
            Color.white : Color.gray : Color.black);
         picture.drawLine(x,y,x,y);
      }
   }
   long f(Point p1, Point p2, int x, int y) {
      return ((p1.x-x)*(p2.y-y)-(p1.y-y)*(p2.x-x))/13;
   }
}